### ISSUE TITLE
Implement `git hub pull merge` option

### ISSUE THREAD

**Issue by joseph-wakeling-sociomantic** [sociomantic/git-hub#154]
`git-hub` currently allows PRs to be rebased onto master, but not merged into master.  This means that, when a merge is considered necessary, there is no uniform or well-defined means to achieve it.  It also makes it likely that in these circumstances people will just click the Merge button in the GitHub UI, which produces a merge commit with a completely unhelpful commit message.

I'd therefore propose that `git-hub` implement a `pull merge` option which compels the user to create a proper merge commit message, and helps them in doing so.  A pattern I have followed, which may be useful here, is:

```
Merge projectowner/projectname#NNN: This is the pull-request title

Here the person authoring the merge commit may write a more
detailed description of the patchset.
```

where `NNN` is the PR number.  The pull-request title should be used by default, but obviously the author of the merge-commit may edit it freely.

**Comment by mihails-strasuns-sociomantic** [comment]
Request from my side - this command should also do rebase _before_ doing the merge and use `--no-ff` flag after, so that even with merge commits history will be purely incremental.

**Comment by joseph-wakeling-sociomantic** [comment]
@mihails-strasuns-sociomantic no, that defeats part of the point of a merge, which is that you get _the same commits_ in the source branch and the branch into which they are merged.

I don't mind if there is a `--with-rebase` option in there to allow the use-case you have in mind, but a merge command should by default reflect normal merge behaviour.

**Comment by mihails-strasuns-sociomantic** [comment]
Than I am totally against such feature and very worried someone does that in their project.

**Comment by joseph-wakeling-sociomantic** [comment]
@mihails-strasuns-sociomantic why?  What you describe as sane behaviour totally gets in the way of, for example, merging a point-release branch sanely into master.

**Comment by mihails-strasuns-sociomantic** [comment]
`git-hub` is about GitHub interaction (i.e. pull requests) and managing release branches can (and should) be done with plain `git` commands. For merging point release branch  `git merge --no-ff` works pretty much as needed.

**Comment by joseph-wakeling-sociomantic** [comment]
What I'm asking here is that the two different options, `merge` and `rebase`, should idiomatically match their counterparts in basic git commands.  Having something different because someone thinks it's a better model is potentially a nasty source of confusion.

If you as a project maintainer want to impose a particular way of doing things on your project, that's fine and you should have that option, but it's simply a case of using it.  That's not hard: it can be a matter of setting a config variable for the repo, or just always using the `--with-rebase` option that I propose.

What I _don't_ think is acceptable is that the default version of a `merge` command should actually rewrite commit history.  That's exactly what a merge is supposed to _avoid_.

**Comment by mihails-strasuns-sociomantic** [comment]
And what I mean is that plain `merge` (in a sense defined by git) must _never_ be used for pull request merging, there can be no legitimate justification for that. Currently it is considered justified to do merge commits to aggregate bunch of refactorings but it is only used as aggregation with empty actual merge commit, rebasing still takes place before. At least that was the intention - I have no idea what people actually do in projects.

It is fine (and even necessary) to do branch merges for version management but that is not `git hub` job.

**Comment by joseph-wakeling-sociomantic** [comment]
> And what I mean is that plain merge (in a sense defined by git) must never be used for pull request merging, there can be no legitimate justification for that.

What on earth do you mean, no legitimate justification?  It's a completely normal means of managing code submissions.  The fact that you may prefer not to do it for some (or even all) projects that you have responsibility for isn't a justification for banning the tool from supporting this feature (or for redefining "merge" to mean "rebase-then-merge").

**Comment by mihails-strasuns-sociomantic** [comment]
"common" != "normal" != "effective"

It is not simply matter of aesthetical preferences. My practical experience from using merges for everything is roughly "spend 1 or 2 full days each month trying to resolve conflicts and get all release branches right" and "convince git to not destroy merge commit when doing rebase" for liberal mixed model.

It mostly works when you only have one master branch and keep merging pull requests there but it is completely not forward-compatible with any more complicated versioning/branching model. Basically the way `git merge` is implemented you need to be extremely careful with merging existing merge commits with actual changesets. Thus using them for pull reuqests makes it much harder to use them for anything else (i.e. point release branch merging). Purity of unmodified hashes is simply not worth such headache for maintenance.

**Comment by joseph-wakeling-sociomantic** [comment]
I think it's important to distinguish between what a tool can do, and how we may wish to use it.  Merges and rebases are two different ways of incorporating new patches into a project.  They both have idiomatic meanings, and they both have advantages and disadvantages.  A good tool should not distort or block those choices.

I don't necessarily disagree with anything you say about good practice.  What I do disagree with is the suggestion that the `git-hub` tool should enforce preferences about "good practice" over respecting the idiomatic meanings of `merge` and `rebase` and the choice of the user as to which is the desirable way to combine those different options.

A good tool should _serve_ project management choices, not impose them.

**Comment by joseph-wakeling-sociomantic** [comment]
FWIW, my long-term preference would be to rename `git-hub`'s `pull` command to `pr` so we could have:

* `git hub pr pull`

* `git hug pr rebase`

* `git hub pr merge`

... as the user wishes.  And then perhaps a generic

* `git hub pr accept`

... which would apply a default means of incorporating the PR into the target branch.  I would be _entirely_ happy with `accept` being defined as "rebase then merge" (or just "rebase" in case of a single commit).
